README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages)
-->

# xen_popup_card

Stunnigly beautiful yet simple to use and customizeable popup card for flutter.

To see an example:
```
cd example/
flutter run --release
```

To use package:
```
import 'package:xen_popup_card/xen_card.dart';
```

### Understanding with image
![alt text](https://raw.githubusercontent.com/xenSlayer/xen_popup_card/master/visual-understanding.png)

### [Example] Screenshot for card with gutter only
![alt text](https://raw.githubusercontent.com/xenSlayer/xen_popup_card/master/with-gutter%20only.png)

### [Example] Screenshot for card with both gutter and appbar
![alt text](https://raw.githubusercontent.com/xenSlayer/xen_popup_card/master/with-gutter-and-appbar.png)

### Gutter
```
XenCardGutter gutter = const XenCardGutter(
  child: Padding(
    padding: EdgeInsets.all(8.0),
    child: CustomButton(text: "close"),
  ),
);
```

### App Bar
```
XenCardAppBar appBar = const XenCardAppBar(
  child: Text(
    "app bar",
    style: TextStyle(fontWeight: FontWeight.bold, fontSize: 28),
  ),
  shadow: BoxShadow(color: Colors.transparent),   // To remove shadow from appbar
);
```

### XenPopupCard
```
XenPopupCard card = XenPopupCard(
  appBar: appBar,
  gutter: gutter,
  body: ListView(
    children: const [
      Text("body"),
    ],
  ),
);
```

### [Example] popup card with only gutter
```
// card with gutter only
// no appbar
Widget cardWithGutterOnly() => TextButton(
      onPressed: () => showDialog(
        context: context,
        builder: (builder) => XenPopupCard(
          gutter: gutter,
          body: ListView(
            children: const [
              Text("body"),
            ],
          ),
        ),
      ),
      child: const Text("open card with  gutter only"),
    );

```

### [Example] popup card with only appbar
```
// card with appbar only
// no gutter
Widget cardWithAppBarOnly() => TextButton(
      onPressed: () => showDialog(
        context: context,
        builder: (builder) => XenPopupCard(
          appBar: appBar,
          body: ListView(
            children: const [
              Text("body"),
            ],
          ),
        ),
      ),
      child: const Text("open card with appbar only"),
    );

```

### [Example] popup card with only body
```
// card with body only
// no appbar and gutter
Widget cardWithBodyOnly() => TextButton(
      onPressed: () => showDialog(
        context: context,
        builder: (builder) => XenPopupCard(
          body: ListView(
            children: const [
              Text("body"),
            ],
          ),
        ),
      ),
      child: const Text("open card with body only"),
    );
```

### [Example] popup card with both gutter and appbar
```
// card with both gutter and app bar
Widget cardWithGutterAndAppBar() => TextButton(
      onPressed: () => showDialog(
        context: context,
        builder: (builder) => XenPopupCard(
          appBar: appBar,
          gutter: gutter,
          body: ListView(
            children: const [
              Text("body"),
            ],
          ),
        ),
      ),
      child: const Text("open card with gutter and app bar"),
    );

```

### Full working example

```
import 'package:flutter/material.dart';
import 'package:xen_popup_card/xen_card.dart';

void main() {
  runApp(const MaterialApp(home: ExampleApp()));
}

class ExampleApp extends StatefulWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  State<ExampleApp> createState() => _ExampleAppState();
}

class _ExampleAppState extends State<ExampleApp> {
  // APP BAR
  //
  // [XenCardAppBar]
  XenCardAppBar appBar = const XenCardAppBar(
    child: Text(
      "app bar",
      style: TextStyle(fontWeight: FontWeight.bold, fontSize: 28),
    ),
    // To remove shadow from appbar
    shadow: BoxShadow(color: Colors.transparent),
  );

  // GUTTER
  //
  // [XenCardGutter]
  XenCardGutter gutter = const XenCardGutter(
    child: Padding(
      padding: EdgeInsets.all(8.0),
      child: CustomButton(text: "close"),
    ),
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: <Widget>[
            cardWithGutterAndAppBar(),
            cardWithBodyOnly(),
            cardWithAppBarOnly(),
            cardWithGutterOnly(),
          ],
        ),
      ),
    );
  }

  // card with both gutter and app bar
  Widget cardWithGutterAndAppBar() => TextButton(
        onPressed: () => showDialog(
          context: context,
          builder: (builder) => XenPopupCard(
            appBar: appBar,
            gutter: gutter,
            body: ListView(
              children: const [
                Text("body"),
              ],
            ),
          ),
        ),
        child: const Text("open card with gutter and app bar"),
      );

  // card with body only
  // no appbar and gutter
  Widget cardWithBodyOnly() => TextButton(
        onPressed: () => showDialog(
          context: context,
          builder: (builder) => XenPopupCard(
            body: ListView(
              children: const [
                Text("body"),
              ],
            ),
          ),
        ),
        child: const Text("open card with body only"),
      );

  // card with appbar only
  // no gutter
  Widget cardWithAppBarOnly() => TextButton(
        onPressed: () => showDialog(
          context: context,
          builder: (builder) => XenPopupCard(
            appBar: appBar,
            body: ListView(
              children: const [
                Text("body"),
              ],
            ),
          ),
        ),
        child: const Text("open card with appbar only"),
      );

  // card with gutter only
  // no appbar
  Widget cardWithGutterOnly() => TextButton(
        onPressed: () => showDialog(
          context: context,
          builder: (builder) => XenPopupCard(
            gutter: gutter,
            body: ListView(
              children: const [
                Text("body"),
              ],
            ),
          ),
        ),
        child: const Text("open card with  gutter only"),
      );
}

// custom button
// ignore
class CustomButton extends StatelessWidget {
  const CustomButton({Key? key, required this.text, this.color})
      : super(key: key);

  final String text;
  final Color? color;

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: () => Navigator.pop(context),
      child: Material(
        borderRadius: BorderRadius.circular(5),
        color: const Color(0xff6200ee),
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: FittedBox(
            fit: BoxFit.scaleDown,
            child: Center(
              child: Text(text,
                  style: const TextStyle(
                      fontSize: 17,
                      color: Colors.white,
                      fontWeight: FontWeight.w500)),
            ),
          ),
        ),
      ),
    );
  }
}

```


CHANGELOG.md
## 0.0.1

* Initial release

## 0.0.2+1

* minor changes


## 0.0.3+2

* 🔨 Fixed padding


lib/src/card.dart
import 'package:flutter/material.dart';

import 'package:xen_popup_card/src/appbar.dart';
import 'package:xen_popup_card/src/gutter.dart';

class XenPopupCard extends StatelessWidget {
  const XenPopupCard({
    required this.body,
    Key? key,
    this.padding,
    this.appBar,
    this.gutter,
    this.cardBgColor,
    this.borderRadius,
  }) : super(key: key);

  /// The main part of the card
  ///
  /// [Widget] body: required
  ///
  /// example: ListView, Container
  ///
  final Widget body;

  /// The padding for body
  ///
  /// `default` : 20
  ///
  final double? padding;

  /// The top part of the card
  ///
  /// `default` : SizedBox()
  ///
  /// [XenCardAppBar] : To create a default card app bar
  ///
  /// Useful if you want to put title or navigation bar
  ///
  /// Alternatively extend the class [XenCardAppBar] to customize
  /// the app bar according to your requirement
  ///
  final XenCardAppBar? appBar;

  /// The bottom part of the card
  ///
  /// `default` : XenCardGutter()
  ///
  /// [XenCardGutter] : To customize default gutter
  ///
  final XenCardGutter? gutter;

  /// The border radius of the card
  ///
  /// `default` : 10
  ///
  /// Alternately you can set border by using [XenCardGutter]
  /// to set bottom border and [XenCardAppBar] to set top border
  ///
  final double? borderRadius;

  /// The background color of the card
  ///
  /// `default` : Colors.white
  ///
  final Color? cardBgColor;

  @override
  Widget build(BuildContext context) {
    final size = MediaQuery.of(context).size;
    return Container(
      // margin
      margin: EdgeInsets.only(
        left: size.width * 0.15 / 2,
        right: size.width * 0.15 / 2,
        top: size.height * 0.1,
        bottom: size.height * 0.1,
      ),
      // material with border radius
      child: Material(
        borderRadius: BorderRadius.circular(borderRadius ?? 20),
        color: cardBgColor ?? Colors.white,
        child: Stack(
          children: [
            // body
            Padding(
              padding: EdgeInsets.only(
                top: appBar == null ? 20 : 80,
                bottom: padding ?? 20,
                left: padding ?? 20,
                right: padding ?? 20,
              ),
              child: body,
            ),
            // appbar
            Align(
              alignment: Alignment.topCenter,
              child: appBar ?? const SizedBox(),
            ),
            // gutter
            Align(
              alignment: Alignment.bottomCenter,
              child: gutter ?? const SizedBox(),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/appbar.dart
import 'package:flutter/material.dart';

///
/// Xen Card App bar
///
class XenCardAppBar extends StatelessWidget {
  const XenCardAppBar({
    required this.child,
    Key? key,
    this.shadow,
    this.padding,
    this.color,
    this.borderRadius,
  }) : super(key: key);

  /// The main widget of the appbar
  ///
  /// [Widget] child: required
  ///
  /// example: Text, Container
  ///
  final Widget child;

  ///
  /// BoxShadow(
  /// color: Colors.black.withOpacity(0.1),
  /// spreadRadius: 1,
  /// blurRadius: 2,
  /// offset: const Offset(1, 2)),
  ///
  /// use BoxShadow(color: Colors.transparent) to remove shadow and line
  final BoxShadow? shadow;

  /// `default` : const EdgeInsets.all(10)
  ///
  /// padding for gutter widget
  final EdgeInsetsGeometry? padding;

  /// `default` : Colors.white
  ///
  /// use to change color of appbar
  final Color? color;

  /// `default` :  10
  ///
  /// Use to change border radius of top corners
  ///
  /// `Warning` : will require changing the gutter borderRadius
  /// to have same top and bottom border radius
  final double? borderRadius;

  @override
  Widget build(BuildContext context) {
    // size
    final size = MediaQuery.sizeOf(context);

    return Container(
      // padding
      padding: padding ?? const EdgeInsets.all(20),
      // border radius
      decoration: BoxDecoration(
        color: color ?? Colors.white,
        borderRadius: BorderRadius.only(
          topLeft: Radius.circular(borderRadius ?? 10),
          topRight: Radius.circular(borderRadius ?? 10),
        ),
        boxShadow: [
          shadow ??
              BoxShadow(
                color: Colors.black.withOpacity(0.1),
                spreadRadius: 1,
                blurRadius: 2,
                offset: const Offset(1, 2),
              ),
        ],
      ),
      // appbar
      child: SizedBox(width: size.width, child: Container(child: child)),
    );
  }
}


lib/src/gutter.dart
import 'package:flutter/material.dart';

/// Xen Card Gutter
class XenCardGutter extends StatelessWidget {
  const XenCardGutter({
    required this.child,
    Key? key,
    this.shadow,
    this.borderRadius,
  }) : super(key: key);

  /// The main widget of the gutter
  ///
  /// [Widget] child: required
  ///
  /// example: Button
  ///
  final Widget child;

  /// `default` :
  ///
  /// BoxShadow(
  /// color: Colors.black.withOpacity(0.1),
  /// spreadRadius: 1,
  /// blurRadius: 2,
  /// offset: const Offset(1, 2)),
  ///
  /// `tip` : use BoxShadow(color: Colors.transparent) to remove shadow and line
  final BoxShadow? shadow;

  /// `default` :  10
  ///
  /// Use to change border radius of bottom corners
  ///
  /// `warning` : will require changing the appbar borderRadius
  /// to have same top and bottom border radius
  final double? borderRadius;

  @override
  Widget build(BuildContext context) {
    final size = MediaQuery.sizeOf(context);

    return Container(
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.only(
          bottomLeft: Radius.circular(borderRadius ?? 10),
          bottomRight: Radius.circular(borderRadius ?? 10),
        ),
        boxShadow: [
          shadow ??
              BoxShadow(
                color: Colors.black.withOpacity(0.1),
                spreadRadius: 1,
                blurRadius: 2,
                offset: const Offset(1, -2),
              ),
        ],
      ),
      // gutter
      child: SizedBox(width: size.width, child: child),
    );
  }
}


example/lib/ui/button.dart
import 'package:flutter/material.dart';

// custom button
// ignore
class CustomButton extends StatelessWidget {
  const CustomButton({Key? key, required this.text, this.color})
      : super(key: key);

  final String text;
  final Color? color;

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: () => Navigator.pop(context),
      child: Material(
        borderRadius: BorderRadius.circular(5),
        color: const Color(0xff6200ee),
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: FittedBox(
            fit: BoxFit.scaleDown,
            child: Center(
              child: Text(text,
                  style: const TextStyle(
                      fontSize: 17,
                      color: Colors.white,
                      fontWeight: FontWeight.w500)),
            ),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:example/ui/button.dart';
import 'package:flutter/material.dart';
import 'package:xen_popup_card/xen_popup_card.dart';

void main() {
  runApp(const MaterialApp(home: ExampleApp()));
}

class ExampleApp extends StatefulWidget {
  const ExampleApp({Key? key}) : super(key: key);

  @override
  State<ExampleApp> createState() => _ExampleAppState();
}

class _ExampleAppState extends State<ExampleApp> {
  // APP BAR
  //
  // [XenCardAppBar]
  XenCardAppBar appBar = const XenCardAppBar(
    child: Text(
      "app bar",
      style: TextStyle(fontWeight: FontWeight.bold, fontSize: 28),
    ),
    // To remove shadow from appbar
    shadow: BoxShadow(color: Colors.transparent),
  );

  // GUTTER
  //
  // [XenCardGutter]
  XenCardGutter gutter = const XenCardGutter(
    child: Padding(
      padding: EdgeInsets.all(8.0),
      child: CustomButton(text: "close"),
    ),
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: <Widget>[
            cardWithGutterAndAppBar(),
            cardWithBodyOnly(),
            cardWithAppBarOnly(),
            cardWithGutterOnly(),
          ],
        ),
      ),
    );
  }

  // card with both gutter and app bar
  Widget cardWithGutterAndAppBar() => TextButton(
        onPressed: () => showDialog(
          context: context,
          builder: (builder) => XenPopupCard(
            appBar: appBar,
            gutter: gutter,
            body: ListView(
              children: const [
                Text("body"),
              ],
            ),
          ),
        ),
        child: const Text("open card with gutter and app bar"),
      );

  // card with body only
  // no appbar and gutter
  Widget cardWithBodyOnly() => TextButton(
        onPressed: () => showDialog(
          context: context,
          builder: (builder) => XenPopupCard(
            body: ListView(
              children: const [
                Text("body"),
              ],
            ),
          ),
        ),
        child: const Text("open card with body only"),
      );

  // card with appbar only
  // no gutter
  Widget cardWithAppBarOnly() => TextButton(
        onPressed: () => showDialog(
          context: context,
          builder: (builder) => XenPopupCard(
            appBar: appBar,
            body: ListView(
              children: const [
                Text("body"),
              ],
            ),
          ),
        ),
        child: const Text("open card with appbar only"),
      );

  // card with gutter only
  // no appbar
  Widget cardWithGutterOnly() => TextButton(
        onPressed: () => showDialog(
          context: context,
          builder: (builder) => XenPopupCard(
            gutter: gutter,
            body: ListView(
              children: const [
                Text("body"),
              ],
            ),
          ),
        ),
        child: const Text("open card with  gutter only"),
      );
}
